grakn-graql/src/main/java/ai/grakn/graql/internal/query/InsertQueryExecutor.java

```diff
@@ -67,285 +63,321 @@
+            the concept represented by the variable `var` is created.
+
+            For example, the property `$x isa $y` depends on the existence of the concept represented by `$y`.
+         */
+
+        Multimap<VarAndProperty, Var> propDependencies =
+                properties.stream().collect(flatteningToMultimap(identity(), VarAndProperty::requiredVars));
+
+        /*
+            `varDependencies.containsEntry(var, prop)` indicates that the concept represented by the variable `var`
+            cannot be created until the property `prop` is inserted.
+
+            For example, the concept represented by `$x` will not exist before the property `$x isa $y` is inserted.
+         */
+        Multimap<Var, VarAndProperty> varDependencies = HashMultimap.create();
+
+        Multimap<VarAndProperty, Var> producedVars =
+                properties.stream().collect(flatteningToMultimap(identity(), VarAndProperty::producedVars));
+
+        Multimaps.invertFrom(producedVars, varDependencies);
+
+        /*
+            Equivalent vars are variables that must represent the same concept as another var.
+
+                 $x label movie, sub entity;
+                 $y label movie;
+                 $z isa $y;
+
+            In this example, `$z isa $y` must not be inserted before `$y` is. However, `$y` does not have enough
+            information to insert on its own! We know `$y` must represent the same concept as `$x`, because they
+            both share the same label property. Therefore, we can share their dependencies, such that:
+
+                varDependencies.containsEntry($x, prop) <=> varDependencies.containsEntry($y, prop)
+
+            Therefore:
+
+                varDependencies.containsEntry($x, `$x sub entity`) => varDependencies.containsEntry($y, `$x sub entity`)
+
+            Now we know that `$y` depends on `$x sub entity` as well as `$x label movie`, which is enough information to
+            insert the type!
+         */
+
+        Partition<Var> equivalentVars = Partition.singletons(Collections.emptyList());
+
+        equivalentProperties(properties).forEach(props -> {
+            // These properties must refer to the same concept, so share their dependencies
+            Collection<VarAndProperty> producers = varDependencies.get(props.first.var());
+            producers.addAll(varDependencies.get(props.second.var()));
+            varDependencies.replaceValues(props.second.var(), Sets.newHashSet(producers));
+
+            equivalentVars.merge(props.first.var(), props.second.var());
+        });
+
+        /*
+            Together, `propDependencies` and `varDependencies` can be composed into a single many-to-many relation:
+
+                dependencies = propDependencies ∘ varDependencies
+
+            By doing so, we map _directly_ between properties, skipping the vars. For example, if we previously had:
```

Comment:
This following portion really makes clear the purpose of computing equivalentVars. Nice.